### ISSUE TITLE
Schema references resolution success depends on file system/OS

### ISSUE THREAD

**Issue by mpataki** [zolyfarkas/spf4j#42]
This issue is in regards to the `maven-avro-schema-plugin`.

When using schema references, there appears to be a ordering dependency where if schema A depends on B, B must be compiled by the plugin before A. I believe this order is defined by the file system/OS's behavior, and is called [this point in code](https://github.com/zolyfarkas/spf4j/blob/master/spf4j-avro-components/maven-avro-schema-plugin/src/main/java/org/spf4j/maven/plugin/avro/avscp/SchemaCompileMojo.java#L631). Here, if the OS decides to list schema file A before B, the plugin will complain with an error like:

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 3.406 s
[INFO] Finished at: 2020-04-07T08:13:41-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.spf4j:maven-avro-schema-plugin:8.8.0:avro-compile (default-avro-compile) on project avro-schemas: Execution default-avro-compile of goal org.spf4j:maven-avro-schema-plugin:8.8.0:avro-compile failed: "EventMetadata" is not a defined name. The type of the "metadata" field must be a defined name or a {"type": ...} expression. -> [Help 1]
```

In this example, `EventMetadata` is schema B from my generalized description above.

Some operating systems will return files ordered alphabetically, some will return them ordered by their created at time, and I'm sure there are other variants. Ideally this plugin would take control over this behavior, thereby keeping it consistent across environments.

The `avro-maven-plugin` also suffers from this same issue, however it allows users to manually configure the order in which schema files will be compiled by the plugin. While this solution isn't the friendliest to the user as they need to think about this order, it is at least a workable solution solution, where currently the `maven-avro-schema-plugin` appears to have no work around available.

Thanks for your good work on this project and taking the time to read this issue :raised_hands:

**Comment by zolyfarkas** [comment]
Thanks for reporting the issue, 

If the underlying issue is ref resolution within the project, I believe you can work around this with the plugin config:
      <useSchemaReferencesForAvsc>${avro.useSchemaReferencesForAvsc</useSchemaReferencesForAvsc>

**Comment by mpataki** [comment]
Hi @zolyfarkas, thanks for the quick response.

I've tried flipping this flag but unfortunately I still get the same issue.

If it would help, I could publish a same project that suffers from this issue.

**Comment by zolyfarkas** [comment]
Any extra info would help, especially if the flag does not help... thanks in advance!

**Comment by mpataki** [comment]
Here's an example repo: https://github.com/mpataki/avro-schema-ref-demo

**Comment by mpataki** [comment]
schema [A](https://github.com/mpataki/avro-schema-ref-demo/blob/master/src/main/resources/avro/01_A.avsc) makes reference to [B](https://github.com/mpataki/avro-schema-ref-demo/blob/master/src/main/resources/avro/01_B.avsc), but depending on how you name schema B, the plugin will produce the error above.

For example, if I rename B's file to `02_B.avsc`, the issue will occur. In putting this repo together I'm realizing that the exact nature of the issue is still unknown to me, because naming the file `00_B.avsc` also produces the error, where I would have expected it to work file if the issue was around lexicographic ordering of files.

**Comment by zolyfarkas** [comment]
thanks, helpful, now I understand the issue you are facing...
avsc was designed to be self contained (no rreferences)... so basically the workaround for this case is to:
   * declare the 2 schemas in one avsc file.
   * use avdl or avpr to declare them. (both have support for import)

But i think there should be a way to make your use case work, without hacks like relying on alphabetical order, etc...

**Comment by mpataki** [comment]
Thanks, this is also helpful. 

Yes declaring both schemas in the same file is ultimately what needs to happen for our downstream tooling to work (confluent's schema registry currently expects this). However our avro use case would really benefit from being able to re-use schemas, which is why I'm taking this approach.

I've spent only a small amount of time looking into avdl/avpr, so I think I'll do some more research there in the meantime. :+1:

**Comment by zolyfarkas** [comment]
@mpataki take a look at https://github.com/zolyfarkas/core-schema  it uses avdl to declareonly schema's (no rpc stuff)

**Comment by zolyfarkas** [comment]
@mpataki  I pushed out to maven central, version 8.8.1-SNAPSHOT that supports avsc files referencing types from other files. File system ordering is irrelevant, and it should support circular references as well... the schemas are published correctly expanded.  In any case I still recommend avdl due to its superior readability... as a FYI there is a avsc 2 avdl converter I wrote at: https://github.com/zolyfarkas/spf4j/blob/master/spf4j-avro-components/spf4j-avro/src/main/java/org/spf4j/avro/schema/SchemaUtils.java#L176

Let me know if this works or not.
